### PR TITLE
Nemotron Conversion script

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        default=None,
+        required=True,
+        help="Path to .nemo file or extracted folder",
+    )
+    parser.add_argument("--output_path", type=str, default=None, required=False, help="Path to HF .bin file")
+    parser.add_argument(
+        "--hf_input_path",
+        type=str,
+        default=None,
+        help="A HF model path, " "e.g. a folder containing https://huggingface.co/nvidia/nemotron-3-8b-base-4k-hf",
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:68]: nemotron 3 on HF is a .nemo model. Maybe we want to use [Minitron-8B-Base](https://huggingface.co/nvidia/Minitron-8B-Base)?

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:1]: 2024

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        "--cpu-only",
+        action="store_true",
+        help="Load model in cpu only. Useful if the model cannot fit in GPU memory, "
+        "but this option makes the conversion script significantly slower.",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def convert_hf_config(nemo_config, tokenizer, vocab_size, dtype, hf_output_path, hf_url="nvidia/nemotron3-8b-base"):
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:93]:
```suggestion
def convert_hf_config(nemo_config, tokenizer, vocab_size, dtype, hf_output_path, hf_url="nvidia/Minitron-8B-Base"):
```

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        nemo_tokenizer.tokenizer.save_pretrained(output_hf_path)
+        logging.info(f"HF AutoTokenizer has been saved to {output_hf_path}")
+    else:
+        raise ValueError(f"Unsupported tokenizer type: library: {tokenizer_cfg.library}, type: {tokenizer_cfg.type}")
+
+
+if __name__ == "__main__":
+    args = get_args()
+    if not args.hf_output_path:
+        assert args.output_path is not None, "Need to provide either provide output_path or hf_output_path"
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:334]:
```suggestion
        assert args.output_path is not None, "Need to provide either output_path or hf_output_path"
```

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    return args
+
+
+def convert_hf_config(nemo_config, tokenizer, vocab_size, dtype, hf_output_path, hf_url="nvidia/nemotron3-8b-base"):
+    """
+    Convert NeMo config to HF config
+    """
+    NEMO_ACT2HF = {
+        "squared-relu": "relu2",
+        "fast-swiglu": "silu",
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:99]: swiglu = silu + gated activation, while squared-relu doesn't have gated activation. Is this reflected anywhere in the config?

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+            archive.close()
+            os.rename(f"{output_hf_path}/{tokenizer_fn}", output_tokenizer)
+        elif os.path.isdir(nemo_file):
+            shutil.copy(f"{nemo_file}/{tokenizer_fn}", output_tokenizer)
+        # We use LlamaTokenizer for sentencepiece based tokenizer
+        tokenizer = LlamaTokenizer.from_pretrained(output_hf_path)
+        # Convert the LlamaTokenizer to a PreTrainedTokenizerFast instance
+        tokenizer = PreTrainedTokenizerFast(
+            tokenizer_object=LlamaConverter(tokenizer).converted(), model_input_names=["input_ids", "token_type_ids"]
+        )
```
Comment (cuichenx) [inline scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:318]: could you add some explanation why we use llama tokenizer class?

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py (hunk 5 above)
Comment (suiyoubi) [reply scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:99]: currently all nemotron models do not use gated activation. The conversion script can handle it in `convert()` but the HF modeling class don't handle it right now. 
We'll handle that once we have those models later if needed

File: scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py (hunk 6 above)
Comment (suiyoubi) [reply scripts/checkpoint_converters/convert_nemotron_nemo_to_hf.py:318]: Added a comment at L307

**Review by cuichenx — APPROVED** [review]
LGTM

**Review by cuichenx — APPROVED** [review]

**Pull request merged**